### PR TITLE
Added http traffic parser

### PR DESCRIPTION
**Pull request by DoomTaper**

### CODE REVIEW

File: ptp/libptp/parser.py [hunk 1]
```diff
@@ -11,6 +11,7 @@
 import fnmatch
 
 from lxml import etree
+import simplejson as json
```
Comment (delta24) [inline ptp/libptp/parser.py]: Use the in built `json` for now - there are lots of dependencies as it is.

File: ptp/libptp/parser.py [hunk 2]
```diff
@@ -274,3 +275,42 @@ def handle_file(cls, pathname='./', filename='*', skip_empty=True, first=True):
...
+class JSONParser(AbstractParser):
+
+    """Specialized parser for JSON files.
+
+    Define the special :func:`handle_file` function in order to process the JSON report file.
+
+    """
+    __format__ = 'json'
+
+    def __init__(self, pathname='./', filename='*.json', first=True):
```
Comment (delta24) [inline ptp/libptp/parser.py]: for the `pathname` : why not use os.path to pass the current folder?

File: ptp/libptp/parser.py [hunk 3]
```diff
@@ -274,3 +275,42 @@ def handle_file(cls, pathname='./', filename='*', skip_empty=True, first=True):
...
+        :raises IOError: when the report file cannot be found.
+        :raises TypeError: when the report file has not the right extension.
+        :raises :class:`simplejson.scanner.JSONDecodeError`: when simplejson cannot parse the JSON file.
+
+        """
+        fullpath = cls._recursive_find(pathname, filename, first=first)
+        if not len(fullpath):
+            raise IOError("Report matching '%s' cannot be found." % filename)
+        fullpath = fullpath[0]
+        data = open(fullpath, 'r')
```
Comment (delta24) [inline ptp/libptp/parser.py]: Use `with open`.

File: ptp/tools/arachni/parser.py [hunk 4]
```diff
@@ -82,23 +84,47 @@ def parse_metadata(self):
...
+        return data
+
+    def parse_report(self, full_parse):
         """Parse the results of the report.
 
         :return: List of dicts where each one represents a discovery.
         :rtype: :class:`list`
 
         """
         self.vulns = [
```
Comment (delta24) [inline ptp/tools/arachni/parser.py]: Does this fit in 120 chars (1 line)?

File: ptp/tools/hoppy/parser.py [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+
+class HoppyParser(FileParser):
+    """Hoppy specialized parser."""
+
+    __tool__ = 'hoppy'
+    __format__ = 'text'
+    __httpfile_format__ = "*.spider"
+    __version__ = (
+        r'(1\.8\.\d+)') # TODO: check for other version, checked only for 1.8.1
```
Comment (delta24) [inline ptp/tools/hoppy/parser.py]: Single line.

File: ptp/tools/skipfish/parser.py [hunk 6]
```diff
@@ -130,7 +135,38 @@ def parse_metadata(self):
...
+        """ Retrieve list of directories found in samples.js file. From all the directories, it reads request.dat
+        and response.dat file and return a list of dict resquests and responses.
+
+        :raises: IOError -- specifying in which directory it doesn't find request.dat/response.dat file and
+        default its value to NOT_FOUND
+        """
+        data = []
+        for dirs in dir_list:
+            try:
+                req_data = open(dirs['dir']+'/request.dat', 'r')
```
Comment (delta24) [inline ptp/tools/skipfish/parser.py]: Ditto - use `with open` to close the file after reading data from it.

File: ptp/tools/w3af/parser.py [hunk 7]
```diff
@@ -101,7 +109,29 @@ def parse_metadata(self):
...
+            response = self._re_response.findall(transaction)[0] + '\n'
+            response_header = self._re_reponse_header.search(response).group()
+            response_status_code = self._re_response_status_code.search(response_header).group()
+            response_body = self._re_response_body.search(response).group(2)
+            data.append({
+                'request': self._re_request.findall(transaction),
+                'response_status_code': response_status_code,
+                'response_header': response_header,
+                'response_body': response_body               
+                })
```
Comment (delta24) [inline ptp/tools/w3af/parser.py]: This should be less indented.

File: ptp/tools/w3af/parser.py [hunk 8]
```diff
@@ -111,4 +141,12 @@ def parse_report(self):
         self.vulns = [
             {'ranking': self.RANKING_SCALE[vuln.get('severity')]}
             for vuln in self.stream.findall('.//vulnerability')]
+        if full_parse:
+            transfile = self._recursive_find(self.pathname, '*.http.txt')
+            if len(transfile):
+                self.raw_transdata = open(str(transfile[0]), 'r').read()
```
Comment (delta24) [inline ptp/tools/w3af/parser.py]: Same here - use `with open` to read and close the file after reading data.

File: ptp/libptp/parser.py (hunk 2 above)
Comment (DoomTaper) [inline ptp/libptp/parser.py]: you mean os.getcwd(), right?

File: ptp/tools/arachni/parser.py (hunk 4 above)
Comment (DoomTaper) [inline ptp/tools/arachni/parser.py]: Yes, it's 112 (pretty close :) ).

**Comment by DePierre** [comment]
@DoomTaper Thank you for the hard work! Although I haven't completed the review of your request, I really appreciate the work you have provided. I like the new tools you have added to ptp, the documentation for the functions your wrote as well as the comments, and the pep8-style :)

However, before continuing the review of your PR, I have some concerns I would like to discuss with you.

### Arachni XML support dropped?

Why did you drop the [XML support of Arachni](https://github.com/owtf/ptp/pull/15/files#diff-329faea9e46f3af8da40d4c54f4b00ffL38)?

I am trying a little bit to keep some retro-compatibility across versions and I am not sure to understand why you couldn't simply add the JSON support while keeping the XML one.

I would prefer to keep XML and have JSON available in addition.

Seems that there is a missing `$` at the end of the regex [here](https://github.com/owtf/ptp/pull/15/files#diff-02d53e0066b8e78f869f42b8f236d86eR27)

### `full_parse`

I understand that `full_parse` corresponds to what we discussed earlier during the project in order to keep ptp fast for OWTF, while offering a complete parsing for other purposes.

However, I believe that it is not properly implemented. From what I read, passing `full_parse` will force the function to return a tuple instead of a list, where the tuple contain the list of vulnerabilities and the list of data.

But why not simply have everything in the list of vulnerabilities? When you pass the flag, ptp will parse more data that will be stored in the dictionary describing the current vulnerability.

Something like:
```python
vuln = {'ranking': HIGH}
if parse_full:
    vuln['request'] = self.parse_extra(...)
    . . .
```

In the case above, `parse_report` always returns a list of dictionary and we keep consistency across all parsers.

Also, [this feels a little bit wrong](https://github.com/owtf/ptp/pull/15/files#diff-329faea9e46f3af8da40d4c54f4b00ffR73). IMO the fact that a parser supports full_parse should be up to the parser itself and not the ptp wrapper file. For instance, [this feels ugly](https://github.com/owtf/ptp/pull/15/files#diff-329faea9e46f3af8da40d4c54f4b00ffR153).

You haven't commented how and why use `full_parse` and `cumulative` with the ptp wrapper file. As a user, if I don't read the source code of ptp.py, I won't know they exist.


### Why break python3 support?

I understand that `owtf` only supports python2, or that you might be more familiar with python2 development. But it doesn't take much to keep `ptp` working under both version.

For instance, you should know that [`dict.has_key()`](https://github.com/owtf/ptp/pull/15/files#diff-02d53e0066b8e78f869f42b8f236d86eR69) has been removed in python3. From https://docs.python.org/3.1/whatsnew/3.0.html#builtins:

> Removed. dict.has_key() – use the in operator instead.

Furthermore, the unit tests from `ptp` can be run under both python2 and python3, so why not make sure of it by running both?

### Why no unit tests?

**I don't see a single unit test with your PR.** You have modified a lot of codes and added new parsers, new files, etc. You have to provide unit tests along, especially when some of your code only support a specific version of the tool, or when your parser contain like 15 regexes.

### Burp parser

If you only [checked for Burp 1.6.30](https://github.com/owtf/ptp/pull/15/files#diff-88d628e53fccf460d8d940f54628f581R23), why did you write a regex that will also accept other versions? Your comment seems to be in contradiction with your code.

[Shouldn't it raise `IOError`](https://github.com/owtf/ptp/pull/15/files#diff-88d628e53fccf460d8d940f54628f581R50) in case the file doesn't exist? So that the user can know if it failed because it is not supported, or the file doesn't exist. (Related to #14)

[Doesn't it throw an exception if there is no such attribute?](https://github.com/owtf/ptp/pull/15/files#diff-88d628e53fccf460d8d940f54628f581R52)

I am not sure why you had to [move `parse_report` it to another function](https://github.com/owtf/ptp/pull/15/files#diff-88d628e53fccf460d8d940f54628f581R83)?

### Hoppy parser

Ditto regarding the [version regex and your comment](https://github.com/owtf/ptp/pull/15/files#diff-67c9ade2c853a5f7513299e657369394R21).

I am not sure to understand [this comment](https://github.com/owtf/ptp/pull/15/files#diff-67c9ade2c853a5f7513299e657369394R52).

Ditto for [`IOError`]()https://github.com/owtf/ptp/pull/15/files#diff-67c9ade2c853a5f7513299e657369394R55.

Why not use `FileParser` [here](https://github.com/owtf/ptp/pull/15/files#diff-402246113244a8640925f5476a0a2fd9R145)?

### Misc

- [ ] [Import should be organized properly](https://github.com/owtf/ptp/pull/15/files#diff-3df63fdf943b3440dd487e671e5c68a7R14)
- [ ] [Missing newline](https://github.com/owtf/ptp/pull/15/files#diff-3df63fdf943b3440dd487e671e5c68a7R278)
- [ ] [Extra newline](https://github.com/owtf/ptp/pull/15/files#diff-88d628e53fccf460d8d940f54628f581R24)
- [ ] [CamelCase](https://github.com/owtf/ptp/pull/15/files#diff-329faea9e46f3af8da40d4c54f4b00ffR86)
- [ ] [Improper docstring](https://github.com/owtf/ptp/pull/15/files#diff-02d53e0066b8e78f869f42b8f236d86eR97)
- [ ] [Missing surrounding whitespaces](https://github.com/owtf/ptp/pull/15/files#diff-02d53e0066b8e78f869f42b8f236d86eR109)
- [ ] [Why not put your name?](https://github.com/owtf/ptp/pull/15/files#diff-88d628e53fccf460d8d940f54628f581R5) The module author is the one who created the module. In the case Burp's parser, it's your creation so be proud :) Btw, you should add your names on the modules you modified.
- [ ] [Missing whitespace before #](https://github.com/owtf/ptp/pull/15/files#diff-88d628e53fccf460d8d940f54628f581R23)
- [ ] [Delete commented code](https://github.com/owtf/ptp/pull/15/files#diff-e4f442fb5b44f642014793bc3e0655c1R59)
- [ ] [Delete commented code](https://github.com/owtf/ptp/pull/15/files#diff-e4f442fb5b44f642014793bc3e0655c1R191)
- [ ] [Missing capital letter](https://github.com/owtf/ptp/pull/15/files#diff-e4f442fb5b44f642014793bc3e0655c1R197)
- [ ] [Isn't it inefficient to call `eval_js` twice?](https://github.com/owtf/ptp/pull/15/files#diff-e4f442fb5b44f642014793bc3e0655c1R201)
- [ ] I will need to check it further but [your use of js2py](https://github.com/owtf/ptp/pull/15/files#diff-e4f442fb5b44f642014793bc3e0655c1R193) feels even more hackish than my use of literal_eval. How do you feel about it?
- [ ] The API (also it's not really an API) exposes `parse_report` and `parse_metadata` to the user. For other functions, feel free to pick any name and even follow the private naming convention (by prefixing _ to the function name). That is to say that [`parse_http`](https://github.com/owtf/ptp/pull/15/files#diff-402246113244a8640925f5476a0a2fd9R112) doesn't feel properly named.

**Comment by DePierre** [comment]
And you broke the unit tests already written for the code base. Either you update them to correspond to your modification, or you update your code to follow the behavior defined in the unit tests.

```bash
nosetests -v -d --with-coverage --cover-package=ptp
. . .
test_ptp_parse_mock_parser (tests.test_ptp.TestPTP) ... No parser matched `ToolParser((), {})`

ERROR
test_ptp_parse_no_tool (tests.test_ptp.TestPTP) ... No parser matched `ToolParser((), {})`

ok

======================================================================
ERROR: Failure: ImportError (cannot import name ArachniXMLParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/nose/loader.py", line 418, in loadTestsFromName
    addr.filename, addr.module)
  File "/usr/local/lib/python2.7/dist-packages/nose/importer.py", line 47, in importFromPath
    return self.importFromDir(dir_path, fqname)
  File "/usr/local/lib/python2.7/dist-packages/nose/importer.py", line 94, in importFromDir
    mod = load_module(part_fqname, fh, filename, desc)
  File "/home/maya/git-repos/ptp/tests/tools/arachni/test_parser.py", line 9, in <module>
    from ptp.tools.arachni.parser import ArachniXMLParser
ImportError: cannot import name ArachniXMLParser

======================================================================
ERROR: test_parser_skipfish_parse_report (tests.tools.skipfish.test_parser.TestSkipfishJSParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/home/maya/git-repos/ptp/tests/tools/skipfish/test_parser.py", line 41, in test_parser_skipfish_parse_report
    report = my_skipfish.parse_report()
TypeError: parse_report() takes exactly 2 arguments (1 given)

======================================================================
ERROR: test_parser_arachni_xml_parse_report (tests.tools.w3af.test_parser.TestW3AFXMLParser)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/local/lib/python2.7/dist-packages/mock/mock.py", line 1305, in patched
    return func(*args, **keywargs)
  File "/home/maya/git-repos/ptp/tests/tools/w3af/test_parser.py", line 222, in test_parser_arachni_xml_parse_report
    report = my_w3af.parse_report()
TypeError: parse_report() takes exactly 2 arguments (1 given)

======================================================================
ERROR: test_ptp_parse_mock_parser (tests.test_ptp.TestPTP)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/maya/git-repos/ptp/tests/test_ptp.py", line 48, in test_ptp_parse_mock_parser
    vulns = my_ptp.parse()
  File "/home/maya/git-repos/ptp/ptp/ptp.py", line 148, in parse
    raise NotSupportedToolError('This tool is not supported by PTP.')
NotSupportedToolError: This tool is not supported by PTP.
. . .
```

**Comment by delta24** [comment]
Did you resolve all of @DePierre's suggestions? @DoomTaper

**Comment by DoomTaper** [comment]
I am working on it @delta24

**Comment by DePierre** [comment]
@DoomTaper I have merged the PR with 713a8fba1a402203de6ca5c02bd5e14e892ba2c3

There are many aspects I do not like in the updated parsers or in the newly implemented parsers but it can be changed in new commits until the next release. However I do appreciate your work and I am very happy to merge your PR :) (with some modifications).

Regarding the different aspects I do not appreciate:

+ The `parse_http` is very clumsy and I don't understand what you tried to implement (a private static class attribute later modified like an instance attribute) so I modified the feature in the PR
+ The advanced parsing (a.k.a. `full_parse` after my changes) is not what PTP should implement. The idea is that each finding should contain more information like the HTTP request sent by the tool to find it, and not just a bulk-parsing of the HTTP requests that you add in a `transactions` key. The way you implemented it is not very helpful as of now IMO.

Feel free yo have a look at the merge and let me know if you have any comment.

**Pull request closed without merging**